release-22.2: server: retry ambiguous kv/sql operations on node startup

aliher1911 opened the pull request:
In TestClusters with more than 4 nodes, wait for full replication
doesn't take updated config into account.
This commit adds new check that zone config was propagated prior to
making replication state checks.

Epic: none

Release note: None

Backport 2/2 commits from #97710.

/cc @cockroachdb/release

---

Previously, nodes could fail on startup if node itself was needed to restore quorum on ranges. This problem was caused by circuit breaker failing fast on live node when talking back to restarting node instead of waiting for raft operation to run again (and succeed).
This PR adds explicit retries and also adds assertions enabled in race build that would fire if startup reaches dist sender or internal sql executor without wrapping it in startup retry helper.

Release note: None

Fixes #74714

---

Includes fix #100583

Release justification: fixes a stability issue during node restarts

aliher1911 committed:
testutil: zone config propagation wait for test cluster

In TestClusters with more than 4 nodes, wait for full replication
doesn't take updated config into account.
This commit adds new check that zone config was propagated prior to
making replication state checks.

Epic: none

Release note: None

aliher1911 committed:
server: retry ambiguous kv/sql operations on node startup

Previously, nodes could fail on startup if node itself was needed to
restore quorum on ranges. This problem was caused by circuit breaker
failing fast on live node when talking back to restarting node instead
of waiting for raft operation to run again (and succeed).
This PR adds explicit retries and also adds assertions enabled in race
build that would fire if startup reaches dist sender or internal sql
executor without wrapping it in startup retry helper.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/100463)
<!-- Reviewable:end -->

knz committed:
server,util: avoid retry forever during startup upon premature shutdown

We need to have the `RunIdempotentWithRetry` calls abort if the
surrounding server shuts down prematurely. This happens most
frequently in tests that fail for another reason; and also
in multitenancy tests with multiple tenant server side-by-side.

Release note: None

aliher1911 commented:
@tbg this is a backport PR
